sklearn/ensemble/_hist_gradient_boosting/splitting.pyx

```diff
@@ -716,6 +773,179 @@ cdef class Splitter:
+            bin_idx = cat_infos[sorted_cat_idx].bin_idx;
+
+            n_samples_left += feature_hist[bin_idx].count
+            n_samples_right = n_samples - n_samples_left
+
+            if self.hessians_are_constant:
+                sum_hessian_left += feature_hist[bin_idx].count
+            else:
+                sum_hessian_left += feature_hist[bin_idx].sum_hessians
+            sum_hessian_right = sum_hessians - sum_hessian_left
+
+            sum_gradient_left += feature_hist[bin_idx].sum_gradients
+            sum_gradient_right = sum_gradients - sum_gradient_left
+
+            if (n_samples_left < self.min_samples_leaf or
+                sum_hessian_left < self.min_hessian_to_split):
+                continue
+            if (n_samples_right < self.min_samples_leaf or
+                sum_hessian_right < self.min_hessian_to_split):
+                break
+
+            gain = _split_gain(sum_gradient_left, sum_hessian_left,
+                                sum_gradient_right, sum_hessian_right,
+                                loss_current_node, monotonic_cst,
+                                lower_bound, upper_bound,
+                                self.l2_regularization)
+            if gain > best_gain and gain > self.min_gain_to_split:
+                found_better_split = True
+                best_gain = gain
+                best_sorted_thres = sorted_cat_idx
+                best_sum_gradient_left = sum_gradient_left
+                best_sum_hessian_left = sum_hessian_left
+                best_n_samples_left = n_samples_left
+
+
+        if found_better_split:
+            split_info.gain = best_gain
+
+            # split_info.bin_idx is unused for categorical splits: cat_bitset
+            # is used instead and set below
+            split_info.bin_idx = 0
+
+            split_info.sum_gradient_left = best_sum_gradient_left
+            split_info.sum_gradient_right = sum_gradients - best_sum_gradient_left
+            split_info.sum_hessian_left = best_sum_hessian_left
+            split_info.sum_hessian_right = sum_hessians - best_sum_hessian_left
+            split_info.n_samples_left = best_n_samples_left
+            split_info.n_samples_right = n_samples - best_n_samples_left
+
+            # We recompute best values here but it's cheap
+            split_info.value_left = compute_node_value(
+                split_info.sum_gradient_left, split_info.sum_hessian_left,
+                lower_bound, upper_bound, self.l2_regularization)
+
+            split_info.value_right = compute_node_value(
+                split_info.sum_gradient_right, split_info.sum_hessian_right,
+                lower_bound, upper_bound, self.l2_regularization)
+
+            # create bitset with values from best_sorted_thres
+            init_bitset(split_info.cat_bitset)
+
+            for i in range(best_sorted_thres + 1):
+                bin_idx = cat_infos[i].bin_idx
+                set_bitset(split_info.cat_bitset, bin_idx)
```

Comment:
@thomasjpfan it seems that the categories that do not respect `MIN_CAT_SUPPORT` will never be in that bitset and will thus always be mapped to the right child.

Isn't this a problem? This could explain why LightGBM scans in both directions maybe. Could you inverstigate? Thanks!

Reply:
I do not think it is related. Lightgbm seems to prefer "going right by default":

https://github.com/microsoft/LightGBM/blob/2870490073c537eeb3648893257be8692b63148c/include/LightGBM/tree.h#L302-L320

In this case, negative categories and missing values just go right. In this case of splitting, "unconsidered categories go right". For reference, the parameter is called `cat_smooth` in lightgbm.

Reply:
> Lightgbm seems to prefer "going right by default"

I find this surprising because for nans in continuous features, they do scan in both direction to decide whether nans should go left, or right. I find it strange that they wouldn't do the same for categorical features. Maybe there's something we're missing here.

Reply:
I checked LightGBM's code, and I am quite convinced that cat_smooth is indeed the reason why they scan in both direction (and we should do the same)

Consider a case with categories `A B C D`, and let's say that C does not satisfy cat_smooth. Let's  say that the ordering is

`B D A`

In the current state of the PR, we only scan from left to right, and thus we force the C category to be on the right child. The only splits that we can allow are:

- B, DAC
- BD, AC


By also scanning from right to left, LightGBM allows for the additional 2 splits:

- A, BDC
- DA, BC

Note that, indeed, the C category is *always* on the right child. But what matters is in fact what other categories we allow to be on the right child with them.

After all this is pretty much the same logic as with NaNs in the continuous case.
I'll push something soon, with some comments.

Reply:
If you are adding scanning in both directions back again, for reference, it was removed in [`dd4898b` (#18394)](https://github.com/scikit-learn/scikit-learn/pull/18394/commits/dd4898b8298414ded0e2d0238d456a0b792907cb)